Mathlib/CategoryTheory/Sites/RegularExtensive.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2023 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson, Filippo A. E. Nuccio, Riccardo Brasca
+-/
+import Mathlib.CategoryTheory.Sites.Coherent
+/-!
+
+# The Regular and Extensive Coverages
+
+This file defines two coverages on a category `C`.
+
+The first one is called the *regular* coverage and for that to exist, the category `C` must satisfy
+a condition called `Regular C`. This means that effective epimorphisms can be "pulled back". The
+covering sieves of this coverage are generated by presieves consisting of a single effective
+epimorphism.
+
+The second one is called the *extensive* coverage and for that to exist, the category `C` must
+satisfy a condition called `Extensive C`. This means `C` has finite coproducts and that those
+are preserved by pullbacks. The covering sieves of this coverage are generated by presieves
+consisting finitely many arrows that together induce an isomorphism from the coproduct to the
+target.
+
+In `extensive_union_regular_generates_coherent`, we prove that the union of these two coverages
+generates the coherent topology on `C` if `C` is precoherent, extensive and regular.
+
+TODO: figure out under what conditions `Regular` and `Extensive` are implied by `Precoherent` and
+vice versa.
+
+-/
+
+namespace CategoryTheory
+
+open Limits
+
+variable (C : Type*) [Category C]
+
+/--
+The condition `Regular C` is property that effective epis can be "pulled back" along any
+morphism. This is satisfied e.g. by categories that have pullbacks that preserve effective
+epimorphisms (like `Profinite` and `CompHaus`), and categories where every object is projective
+(like  `Stonean`).
+-/
+class Regular : Prop where
+  /--
+  For `X`, `Y`, `Z`, `f`, `g` like in the diagram, where `g` is an effective epi, there exists
+  an object `W`, a morphism `i : W ⟶ Z` and an effective epi `h : W ⟶ X` making the diagram
+  commute.
+  ```
+  W --i-→ Z
+  |       |
+  h       g
+  ↓       ↓
+  X --f-→ Y
+  ```
+  -/
+  exists_fac : ∀ {X Y Z : C} (f : X ⟶ Y) (g : Z ⟶ Y) [EffectiveEpi g],
+    (∃ (W : C) (h : W ⟶ X) (_ : EffectiveEpi h) (i : W ⟶ Z), i ≫ g = h ≫ f)
+
+/--
+Describes the property of having pullbacks of morphsims into a finite coproduct, where one
+of the morphisms is an inclusion map into the coproduct (up to isomorphism).
+-/
+class HasPullbacksOfInclusions : Prop where
+    /-- For any morphism `f : X ⟶ Z`, where `Z` is the coproduct of `i : (a : α) → Y a ⟶ Z` with
+    `α` finite, the pullback of `f` and `i a` exists for every `a : α`. -/
+    has_pullback : ∀ {X Z : C} {α : Type _} (f : X ⟶ Z) {Y : (a : α) → C}
```

Comment:
What universe is this in? Is this the same universe as `C`, or a different universe? I think it would be best to name the universes in this file rather than using `Type*`.